FlowCrypt/Controllers/Compose/ComposeViewController.swift

```diff
@@ -211,57 +222,65 @@ extension ComposeViewController {
 
         showSpinner("sending_title".localized)
 
-        Promise<Bool> { [weak self] in
-            try await(self!.encryptAndSendMessage())
-        }.then(on: .main) { [weak self] sent in
-            if sent { // else it must have shown error to user
-                self?.handleSuccessfullySentMessage()
+        switch prepareMessage() {
+        case .failure(let error):
+            switch error {
+            case .internalError(let message):
+                showAlert(message: message)
+            case .missedSender:
+                showAlert(message: "compose_no_pub_sender".localized)
+            case .noPubKeyForRecipient:
+                // already handled
+                break
             }
-        }.catch(on: .main) { [weak self] error in
-            self?.showAlert(error: error, message: "compose_error".localized)
+        case .success(let data):
+            self.messageSender.sendMail(mime: data)
+                .retry(3)
+                .sink{ [weak self] result in
+                    guard case .failure(let error) = result else {
+                        return
+                    }
+                    self?.showAlert(error: error, message: "compose_error".localized)
+                } receiveValue: { [weak self] _ in
+                    self?.handleSuccessfullySentMessage()
+                }
+                .store(in: &cancellable)
         }
     }
 
-    private func encryptAndSendMessage() -> Promise<Bool> {
-        Promise<Bool> { [weak self] () -> Bool in
-            guard let self = self else { return false }
+    private func prepareMessage() -> Result<Data, SendMessageError> {
```

Comment:
Got it, the futures are similar to promises. Now I understand it it a bit better.

What is this particular method - it returns `Result<Data, SendMessageError>`. How is that different from `Future<...>`?

Reply:
This is return synchronously.
In this case it's just a "helper function"